### PR TITLE
Make Tracer::Close flush the writer

### PR DESCRIPTION
**Pull request by willgittoes-dd**
Also, make writer write-retry wait cancellable.

### CODE REVIEW

File: src/agent_writer.h [hunk 1]
```diff
@@ -34,9 +34,9 @@ class AgentWriter : public Writer {
 
   void write(Trace trace) override;
 
-  // Send all buffered Traces to the destination now. Will block until sending is complete. This
-  // isn't on the main Writer API because real code should not need to call this.
```
Comment (willgittoes-dd) [inline src/agent_writer.h]: I have altered the agreement.

File: src/writer.h [hunk 2]
```diff
@@ -41,6 +46,9 @@ class ExternalWriter : public Writer {
   // Implements Writer methods.
   void write(Trace trace) override;
 
+  // No flush implementation, since ExternalWriter is not in charge of its own writing schedule.
```
Comment (willgittoes-dd) [inline src/writer.h]: Is this true @cgilmour ?

File: src/agent_writer.cpp [hunk 3]
```diff
@@ -137,14 +137,12 @@ bool AgentWriter::retryFiniteOnFail(std::function<bool()> f) const {
       return true;
     }
     {
-      // Just check for stop_writing_ between attempts. No need to allow wake-from-sleep
```
Comment (willgittoes-dd) [inline src/agent_writer.cpp]: I changed my mind.

File: src/writer.h (hunk 2 above)
Comment (cgilmour) [reply src/writer.h]: Yup.

**Pull request merged**